deep-learning/src/main/scala/com/microsoft/azure/synapse/ml/onnx/ONNXUtils.scala

```diff
@@ -0,0 +1,295 @@
+    }
+
+    if (tensorInfo.isScalar) {
+      dataType
+    } else if (tensorInfo.getShape.length == 1) {
+      // first dimension is assumed to be batch size.
+      dataType
+    } else {
+      nestedArrayType(tensorInfo.getShape.length - 1, dataType)
+    }
+  }
+
+  @tailrec
+  private[onnx] def mapValueInfoToDataType(valueInfo: ValueInfo): DataType = {
+    valueInfo match {
+      case mapInfo: MapInfo =>
+        val keyDataType = mapOnnxJavaTypeToDataType(mapInfo.keyType)
+        val valueDataType = mapOnnxJavaTypeToDataType(mapInfo.valueType)
+        MapType(keyDataType, valueDataType)
+      case seqInfo: SequenceInfo =>
+        if (seqInfo.sequenceOfMaps) {
+          mapValueInfoToDataType(seqInfo.mapInfo)
+        } else {
+          mapOnnxJavaTypeToDataType(seqInfo.sequenceType)
+        }
+      case tensorInfo: TensorInfo =>
+        mapTensorInfoToDataType(tensorInfo)
+    }
+  }
+
+  private[onnx] def mapOnnxValueToArray(value: OnnxValue): Seq[Any] = {
+    value.getInfo match {
+      case tensorInfo: TensorInfo =>
+        if (tensorInfo.isScalar)
+          Seq(value.getValue)
+        else {
+          value.getValue.asInstanceOf[Array[_]].toSeq
+        }
+      case sequenceInfo: SequenceInfo =>
+        if (sequenceInfo.sequenceOfMaps) {
+          value.getValue.asInstanceOf[java.util.List[java.util.Map[_, _]]]
+            .asScala.toArray.map(_.asScala.toMap)
+        } else {
+          value.getValue.asInstanceOf[java.util.List[_]].asScala
+        }
+      case _: MapInfo =>
+        Array(value.getValue.asInstanceOf[java.util.Map[_, _]].asScala.toMap)
+    }
+  }
+
+  private[onnx] def createTensor(env: OrtEnvironment, tensorInfo: TensorInfo, batchedValues: Seq[_]): OnnxTensor = {
+    val shape: Array[Long] = tensorInfo.getShape
+    // the first dimension of the shape can be -1 when multiple inputs are allowed. Setting it to the real
+    // input size. Otherwise we cannot create the tensor from the 1D array buffer.
+    if (shape.tail.contains(-1))
+      throw new Exception(s"The input tensor has shape [${shape.mkString}], " +
+        s"but -1 is only allowed at the first dimension (batch size).")
+    shape(0) = batchedValues.length
+
+    val size = shape.product.toInt
+
+    tensorInfo.`type` match {
+      case OnnxJavaType.FLOAT =>
+        val buffer = FloatBuffer.allocate(size)
+        val actualCount = writeNestedSeqToBuffer[Float](batchedValues, buffer.put(_))
+        assertBufferElementsWritten(size, actualCount)
+        buffer.rewind()
+        OnnxTensor.createTensor(env, buffer, shape)
+      case OnnxJavaType.DOUBLE =>
+        val buffer = DoubleBuffer.allocate(size)
+        val actualCount = writeNestedSeqToBuffer[Double](batchedValues, buffer.put(_))
+        assertBufferElementsWritten(size, actualCount)
+        buffer.rewind()
+        OnnxTensor.createTensor(env, buffer, shape)
+      case OnnxJavaType.INT8 =>
+        val buffer = ByteBuffer.allocate(size)
```

Comment:
I assume you cant make this shorter because of the erasure, but you might be able to make the inner part a function so that its less repetition

Reply:
this was just copied, but sure I can refactor it

Reply:
Won't fix. I tried several ideas, but they all are worse than the existing code. The problem is that pretty much every line has type-specific calls that can't be generalized. Everything in the dependencies is hardwired to specific XBuffer classes, rather than an underlying general Buffer interface.
You can pass all the relevant funcs around into 1 general method, but that doesn't shorten the code. Just makes it longer actually. You could combine the assert and the write, but that ends up being ugly code. In the end, I find this more readable.